### ISSUE TITLE
Koboldcpp allows any web site the user browses to to connect to the local API endpoint (localhost:5001)

### ISSUE THREAD

**Issue by RauliRuohonen** [LostRuins/koboldcpp#539]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [X] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

I expected to be the only one to be able to connect to the HTTP endpoint koboldcpp opens, especially when I set the host setting to "localhost".

Regardless of any possible methods of authentication to restrict access to the endpoint, I'd still expect it to be closed to all non-local access unless the user explicitly allows it. Secure by default / defense in depth.

# Current Behavior

Koboldcpp [sets the CORS header](https://github.com/LostRuins/koboldcpp/blob/56a5fa7a60e87ac75a3872fffc163a3ecd7641bf/koboldcpp.py#L851) `access-control-allow-origin: *`, which allows any JavaScript run on any of the user's web browsers to connect to the `localhost:5001` endpoint.

E.g. if you browse to some random web site in Chrome, the following can happen:
1. The web site runs (untrusted) JavaScript code in your browser, as most web sites these days do
2. The JavaScript code  attempts to connect to `localhost:5001`
3. Your browser runs locally and has access to `localhost:5001` regardless of any firewalls or the `host` setting
4. The JavaScript runs in your browser, so it runs locally too, and has access to `localhost:5001` if the browser allows it
5. The browser tries to connect `localhost:5001` and asks the endpoint if it's OK to connect. It does this by checking the CORS header.
6. Koboldcpp responds with the CORS header `access-control-allow-origin: *`, which means "yes, anyone can connect!"
7. The web browser allows connection from the (untrusted) JavaScript code to `localhost:5001`
8. The JavaScript code can also communicate with the remote site it was loaded from. This means that the owner of the remote site can open a connections to your `localhost:5001` as long as you keep the web site open in your browser. The JavaScript code works as a proxy in this scenario.

You can e.g. try visiting [this web site](https://http.jameshfisher.com/2019/05/26/i-can-see-your-local-web-servers/) with koboldcpp running, and it will tell you that it can see `localhost:5001`.

Note also that the current default settings are to listen on all interfaces. Any computer on the local network that can connect to the port 5001 on your machine is allowed access by CORS. This means that any web sites anyone on those computers visits can connect too — their browser is not going to prevent the connection.
```

**Comment by LostRuins** [comment]
You can bind to a specific interface with the `--host` flag, e.g. `--host 192.168.1.177`. If none is specified, it listens on `0.0.0.0`  (all interfaces) by default.

The CORS headers are set for usability - it is intended that different web frontends on different hosts are able to utilize a KoboldCpp instance running on your local machine. If you would prefer to disable them, you may modify the headers in the python code https://github.com/LostRuins/koboldcpp/blob/concedo/koboldcpp.py#L850 to restrict it to your specific use case. Note that CORS is only a *browser* restriction - a simple python script is free to completely ignore CORS rules for example.

Perhaps an alternative solution would be to set inbound firewall rules or iptables restrictions for the port that KoboldCpp is operating on. 

You may also optionally choose to use a different port with the `--port` flag (defaults to 5001). 

Lastly, from a security perspective the risk is low as there are no codepaths which grant read or write access to the local filesystem while the server is running.

**Comment by aleksusklim** [comment]
I realized what is going on here.

I thought that if I'll use the stand-alone browser without internet access to open embedded Lite there locally – I'm safe. Because I have a software firewall that blocks non-whitelisted applications, and neither koboldcpp nor that browser is in exceptions.

But!! My ordinary browser can still access `127.0.0.1:5001`, and because of `Access-Control-Allow-Origin: *` any third-party website I open – could access it too!

This is definitely not what I want. And it is not enough to think that "koboldcpp does not access filesystem" – no, malicious requests should not be allowed, no matter how much of harm they can do.
Development of new features (as recently proposed model reloading at runtime, for example) should not be restricted by the fact that "well, actually the server is not secure and anybody can send any request, so the proposed function is too dangerous in these conditions"!

I see two approaches how to enhance security:
1. Let user to restrict access to the list of trusted domains. This list should be used in both `Access-Control-Allow-Origin` header, and as white-list to check `Origin` header against. Actually, the browser is obligated to send the correct Origin header when making CORS requests, so I cannot see why changing of `*` is also required. But everybody everywhere are telling that `Access-Control-Allow-Origin` must be restricted too, and the sole Origin whitelisting is not enough. (You can hint for a known default domains, like `127.0.0.1, localhost, lite.koboldai.net`)
2. Password authentication. Since we are protecting the unintentional browser access, the password should be checked inside the request payload, and not inside cookies or WWW-Authenticate authorization headers (because that is harder to make it work because of `Access-Control-Allow-Credentials` and other CORS methods that are not so easy to implement reliably and securely right away). Also, password authentication will restrict other LAN clients in setups where there are many users in local network but the server is meant to be private for only a few of them.

You can implement these options in Network tab (and command-line):
- "Restrict access to these domains (comma-separated)": when `*` or empty (default) – do nothing as currently; otherwise you should issue `Access-Control-Allow-Origin` for each domain in the list, and check incoming requests for `Origin` match if it is present (allowing anything without Origin since that is not coming from a browser), responding 403 for unknown origins and printing a warning in console (briefly explaining how to add that particular domain to allowed list).
- "Protect the server with a password": I think the simplest way would be including it into the URL of Lite, for example `?password=123` (or `#password=123`) and then parsing it with JS to send the proper key (that you should document) with every request. This way, a third-party / malicious client won't be able to get the correct URL to parse the password from it. Also you can include a checkbox "Use random passwords each time" which will create a unique long password URL for each server restart (printing it to console and directly opening the browser is configured). Expect other kobold-api clients to support setting the password in their GUIs in future as well (for now this will of course work only for Lite).

What do you think?

**Comment by LostRuins** [comment]
If you want to set up authentication, please bind only to the local interface `--host 127.0.0.1`

Then run a reverse proxy such as https://gitgud.io/khanon/oai-reverse-proxy 

Then you will be able to setup API keys to restrict access to whoever you want.

**Comment by aleksusklim** [comment]
(Technically, this supersedes the password-protection problem, but still leaves "I opened a malicious website in the browser on the same machine", since direct access to 5001 is possible for all localhost processes)

**Comment by LostRuins** [comment]
Run it within a docker container, and only expose the proxied port, not the original koboldcpp port then.

**Comment by henk717** [comment]
The only function the API exposes is AI generation, anything regarding stories in your KoboldAI Lite session remains in your browser and can not be accessed by other websites. So if there was a malicious website the worst they can do is use your GPU to generate a few prompts.

**Comment by noisefloordev** [comment]
Allowing random snoopy websites to see that you're running a Kobold server and which model you're using is a privacy concern by itself, but looking further, the api/extra/generate/check endpoint returns the most recent generation.  Any website you visit can view your most recent conversation, which I think most people would be surprised by.  There are also security concerns about a local server being accessible by every website you visit when it isn't necessary.

I've made a first pass at implementing an access list for Access-Control-Allow-Origin like @aleksusklim suggested above:

https://github.com/noisefloordev/koboldcpp/tree/fix-open-access

It allolws access to localhost and local networks by default, so people running a local frontned should be unaffected.  This is incomplete and only gets the main handling in place, so I haven't opened a PR.  It still needs an option added to add to the allowed list, probably a comma-separated hostname list.  Handling for disallowed hosts is a hack at the moment and just throws an exception, so it doesn't return a 403 like it should.

**Comment by henk717** [comment]
Sent a DM to lostruins with a suggestion on tackling the check endpoints since that is a seperate discussion from this proposal but thanks for bringing that to my attention.

Aa for your idea do understand that a very high number of users uses KCPP on third party websites usually with a remote link. So cross origin should be allowed from any domain when --remote is used and should have an obvious way of unblocking for those not using remote links.

On those conditions I am personally in favor of it but ill let lostruins decide.

**Comment by LostRuins** [comment]
agreed on the `api/extra/generate/check` which I will mandate the genkeys for when running in multiuser mode. 

I think it is a tradeoff between security and practicality when it comes to CORS. There are multiple downstream 3rd party clients or websites that use the KoboldCPP APIs, and many non-technical users may not know how to deal with it. You also have cases where people run a client and connect to another host over their own LAN.

In any case, CORS depends on the compliance of the browser, and will not prevent any other software from accessing the endpoint either. If you really need security, using authentication credentials via a proxy would be the best approach (as mentioned earlier in this thread). 

I might consider adding API key support in future, but the current KAI API spec does not permit this and would have to be extended to avoid disrupting existing frontends.

**Comment by noisefloordev** [comment]
It doesn't require technical knowledge, you'd just enter "aichatfrontend.com" into an "Allow Remote Services" text field in the setup dialog.  My implementation does allow local network IPs by default.

CORS isn't for preventing random malicious software from making API calls.  I'm not worried about that on my own closed-off home network.  I *am* worried about random websites being able to talk to local servers that were never intended to be accessible, just because I clicked a link.  That's something CORS is designed to prevent, and it's been implemented everywhere since around 2010.  (If your browser somehow lets websites make arbitrary fetches without CORS and the other restrictions browsers place on them, you'll have much bigger security problems than this.)

> agreed on the api/extra/generate/check which I will mandate the genkeys for when running in multiuser mode.

I don't quite follow, can you clarify this?  I'm not in multi-user mode, and websites can fetch api/extra/generate/check and read my last generation.

**Comment by LostRuins** [comment]
So to understand this, you need a bit of background on the history of KAI.

The original KoboldAI API only has a sync generate `/api/v1/generate` endpoint - there was no API streaming. There's also no support for authentication.

When KoboldCpp was first created, it adopted that endpoint's schema. 

Subsequently, KoboldCpp implemented polled-streaming in a backwards compatible way. That is - an ongoing sync generation can be polled at `api/extra/generate/check` to get the generation progress. For clients that did not wish to update, they could continue using sync generate as before. This carried on for a while, Lite and a few other clients adopted this new endpoint.

Then even later, KCPP added multiuser mode in backwards compatible way, allowing different requests by multiple clients to be queued. This created a situation where multiple clients could be polling each other's requests progress. To solve this, **genkeys** were added, which is a unique ID a client assigns to a request when making it. Subsequent `api/extra/generate/check` calls would return the status of the request associated with the matching genkey. However, to allow for backwards compatibility with clients that *did not* send genkeys, when the queue was empty, calling `api/extra/generate/check` without a genkey would act as a wildcard and obtain the last progress similar to original behavior.

So that brings us to the present - for the next version I will mandate the genkey parameter when used in multiuser mode, ensuring that nobody without the correct ID can read the request. For single user mode I plan to keep the original behavior, to preserve compatibility with older frontends that do not send genkeys when polling for checking status.

This is all kind of just background info but the TLDR is suffice to say that in future, just run in multiuser mode and `api/extra/generate/check` should not leak any prompts.

**Comment by LostRuins** [comment]
> It doesn't require technical knowledge, you'd just enter "aichatfrontend.com" into an "Allow Remote Services" text field in the setup dialog. 

Don't want to sound overdramatic but I think a layerperson probably doesn't even know what a hostname is. They've learned to proxy KCPP through guides that teach them to tunnel through Ngrok, localtunnel, trycloudflare but may not even have physical access to the machine or launch scripts e.g. running on google colab. Yes I know whitelisting a specific host is not hard, but it does add additional friction to the setup process.

**Comment by aleksusklim** [comment]
You can fake a dummy response like "KoboldCPP is refusing to work at {DOMAIN}; to allow external access for this domain – launch the server with …" to all APIs that return generated text (instead of returning 403).

This way users would see the error right in their interfaces where the generated result should have been appearing.

Also, considering proxy tunnels: if the host can be reasonably checked with a wildcard (for example `*.google.com` or `*.clouflare.com`; I don't know actual ones) – you can whitelist that by default if the option was not passed, giving a moderate security level (by trusting those companies).

Otherwise, the most secure approach is firstly check `Origin:` header against the current whitelist, and if matched – only then issue `Access-Control-Allow-Origin:` only for this exact domain to not give away other allowed domains.

**Comment by LostRuins** [comment]
CORS doesn't work like that. If the preflight fails, the actual request doesn't even get sent by the browser. From the server's perspective they will only see the OPTIONS.

**Comment by aleksusklim** [comment]
I mean, `OPTIONS` still contains `Origin` (as per https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request), so everything I said is true.

For faking response way – yeah, you would have to keep `*` and reply with 200 to anyone.

**Comment by Atom1cByte** [comment]
Was just about to file a security report about this same issue but then saw this. This doesn't seem to be patched yet...

**Comment by LostRuins** [comment]
Why would it need to be patched? API key support has already been added, and also an admin password. Without it they can't access your instance. The expectation is that if you run the service at a certain port you want it to be accessible at that port. If you only wish to bind to localhost you can also do that with `--host 127.0.0.1` which will only make the instance accessible to the local machine.

**Comment by henk717** [comment]
Its not a bug, users want to be able to use our backend with a frontend of choice and need this feature. Sensitive data cant be accessed on the default settings, only GPU resources can which can be optionally restricted with a password to make that impossible.

**Comment by Atom1cByte** [comment]
> If you only wish to bind to localhost you can also do that with --host 127.0.0.1 which will only make the instance accessible to the local machine.

The rest of the argument seems sound, but the point of the bug seems to be that it *is* accessible as long as the local machine visits a website which can puppeteer it to make the local request.

Other than that, though, I agree that the impact isn't very high since all that can be done is use GPU resources.

**Comment by henk717** [comment]
Yes and if that is a concern its possible to specify an API password and then all GPU requests without that password are rejected.

The reason it has to be like it is is so that browser based static pages that wish to use the AI can use the AI. This is the case for our own online koboldai.net website, but also for UI's like mikupad.

As an API server we want the users to be able to access the API server. Malicious scripts targeting koboldcpp are unlikely and we made sure that we dont have any defaults where they could place files on your computer, read your prompts etc.

One notable one for example is the polling endpoint, if polled streaming is used that can fetch the content of the current / last generation. Not only is this mode no longer the default in our UI, there is an automatically generated key that even when it is used protects that endpoint and this key changes every generation. And if polled streaming isn't being used this endpoint isnt usable either.

Its things like that where we ensure our endpoints are secure rather than inaccessible by things the user wants to use.

**Comment by Atom1cByte** [comment]
Super huge props to you guys for keeping security in mind in all this. I agree. It's an edge case and very unlikely, but if you change your mind, you can just implement a host-header check.

Once again, very great aim and the security is better than most.
